core/tests/DesktopTests/AdalHeadlessAutomationTestApp/App/AuthenticationHelper.cs

```diff
@@ -0,0 +1,375 @@
+            {
+                TokenCache.DefaultShared.ReadItems();
+                List<KeyValuePair<TokenCacheKey, AuthenticationResultEx>> CacheItems = QueryCache(input["authority"],
+                    input["client_id"], input["user_identifier"]);
+
+                foreach (KeyValuePair<TokenCacheKey, AuthenticationResultEx> item in CacheItems)
+                {
+                    // if resource was passed to cache lookup, then only expire token for that resource.
+                    // otherwise expire all matching access tokens.
+                    if (input["resource"] == null || item.Key.ResourceEquals(input["resource"]))
+                    {
+                        var updated = item;
+                        updated.Value.Result.ExpiresOn = DateTime.UtcNow - TimeSpan.FromSeconds(1);
+                        UpdateCache(item, updated);
+                    }
+                }
+                Dictionary<string, object> output = new Dictionary<string, object>();
+                //Send back error if userId or displayableId is not sent back to the user
+                output.Add("expired_access_token_count", CacheItems.Count.ToString());
+                return output.FromDictionaryToJson();
+                             
+            });
+
+            return await myTask.ConfigureAwait(false);
+        }
+
+        public static async Task<string> InvalidateRefreshToken(Dictionary<string, string> input)
+        {
+            Dictionary<string, object> output = new Dictionary<string, object>();
+            Task<string> myTask = Task<string>.Factory.StartNew(() =>
+            {
+                try
+                {
+                    TokenCache.DefaultShared.ReadItems();
+                    List<KeyValuePair<TokenCacheKey, AuthenticationResultEx>> CacheItems = QueryCache(input["authority"],
+                    input["client_id"], input["user_identifier"]);
+
+                    foreach (KeyValuePair<TokenCacheKey, AuthenticationResultEx> item in CacheItems)
+                    {
+                        var updated = item;
+                        updated.Value.RefreshToken = "bad_refresh_token";
+                        updated.Value.Result.ExpiresOn = DateTime.UtcNow;
+                        UpdateCache(item, updated);
+                    }
+                    //Send back error if userId or displayableId is not sent back to the user
+                    output.Add("invalidated_refresh_token_count", CacheItems.Count.ToString());                   
+                }
+                catch (Exception exc)
+                {
+                    output.Add("error", exc.Message); 
+                }
+                return output.FromDictionaryToJson();
+            });
+
+            return await myTask.ConfigureAwait(false);
+        }
+
+        public static async Task<string> ReadCache()
+        {
+            Task<string> myTask = Task<string>.Factory.StartNew(() =>
```

Comment:
Does everything in this class need to be async? Why can't ReadCache return a string?